File: Ahorn/entities/smwTrack.jl
module EeveeHelperSMWTrack

using ..Ahorn, Maple

@mapdef Entity "EeveeHelper/SMWTrack" SMWTrack(x::Integer, y::Integer, color::String="FFFFFF", inactiveColor::String="",
    flag::String="", startOpenFlag::String="", endOpenFlag::String="",
    startOpen::Bool=false, endOpen::Bool=false, notFlag::Bool=false, hidden::Bool=false, hideInactive::Bool=false)

const placements = Ahorn.PlacementDict(
    "SMW Track (Eevee Helper)" => Ahorn.EntityPlacement(
        SMWTrack,
        "line"
    )
)

Ahorn.nodeLimits(entity::SMWTrack) = 1, -1

function Ahorn.selection(entity::SMWTrack)
    x, y = Ahorn.position(entity)

    res = Ahorn.Rectangle[Ahorn.Rectangle(x - 4, y - 4, 8, 8)]

    for node in get(entity.data, "nodes", ())
        nx, ny = Int.(node)

        push!(res, Ahorn.Rectangle(nx - 4, ny - 4, 8, 8))
    end

    return res
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::SMWTrack, room::Maple.Room)
    x, y = Ahorn.position(entity)

    nodes = get(entity.data, "nodes", [])
    fullNodes = [(x, y); nodes]

    rawColor = Ahorn.argb32ToRGBATuple(parse(Int, get(entity.data, "color", "ffffff"), base=16))[1:3] ./ 255
    color = (rawColor..., 1.0)

    for i in 1:length(fullNodes)-1
        Ahorn.drawLines(ctx, [fullNodes[i], fullNodes[i+1]], (0.0, 0.0, 0.0, 1.0), thickness=4)
        Ahorn.drawLines(ctx, [fullNodes[i], fullNodes[i+1]], color, thickness=2)
    end

    startOpen = get(entity.data, "startOpen", false)
    endOpen = get(entity.data, "endOpen", false)

    for i in 1:length(fullNodes)
        nx, ny = Int.(fullNodes[i])
        if (i == 1 && !startOpen) || (i == length(fullNodes) && !endOpen)
            Ahorn.drawRectangle(ctx, nx - 4, ny - 4, 8, 8, (0.0, 0.0, 0.0, 1.0))
            Ahorn.drawRectangle(ctx, nx - 3, ny - 3, 6, 6, color)
        else
            Ahorn.drawCircle(ctx, nx, ny, 3, (0.0, 0.0, 0.0, 1.0))
            Ahorn.drawCircle(ctx, nx, ny, 2, color)
        end
    end
end

end

File: Loenn/entities/smwTrack.lua
local utils = require("utils")
local drawableLine = require("structs.drawable_line")
local drawableSprite = require("structs.drawable_sprite")

local black = { 0.0, 0.0, 0.0, 1.0 }
local pointTexture = "objects/EeveeHelper/smwTrack/point"
local endTexture = "objects/EeveeHelper/smwTrack/end"

local function addAll(to, from)
    for _, value in ipairs(from) do
        table.insert(to, value)
    end

    return to
end

local function flattenNodes(nodes)
    local flattened = {}

    for _, node in ipairs(nodes) do
        table.insert(flattened, node.x)
        table.insert(flattened, node.y)
    end

    return flattened
end

local smwTrack = {
    name = "EeveeHelper/SMWTrack",
    depth = 9000 - 10,
    nodeLimits = { 1, -1 },
    nodeLineRenderType = "line",

    placements = {
        name = "default",
        placementType = "line",
        data = {
            color = "FFFFFF",
            inactiveColor = "",
            flag = "",
            startOpenFlag = "",
            endOpenFlag = "",
            startOpen = false,
            endOpen = false,
            notFlag = false,
            hidden = false,
            hideInactive = false,
        }
    },

    fieldInformation = {
        color = {
            fieldType = "color"
        },
        -- once lonn supports empty values being valid for color fields, uncomment this
        --[[
        inactiveColor = {
            fieldType = "color"
        }
        ]]
    },

    selection = function (room, entity)
        local main = utils.rectangle(entity.x - 4, entity.y - 4, 8, 8)
        local nodeSelections = {}
        for _, node in ipairs(entity.nodes) do
            table.insert(nodeSelections, utils.rectangle(node.x - 4, node.y - 4, 8, 8))
        end

        return main, nodeSelections
    end,

    sprite = function (room, entity)
        local sprites = {}

        local startOpen = entity.startOpen
        local endOpen = entity.endOpen
        local fullNodes = addAll({entity}, entity.nodes)
        local color = utils.getColor(entity.color or "ffffff")
        local flattenedNodes = flattenNodes(fullNodes)

        table.insert(sprites, drawableLine.fromPoints(flattenedNodes, black, 3, 1, 1))
        table.insert(sprites, drawableLine.fromPoints(flattenedNodes, color, 1))

        for i, node in ipairs(fullNodes) do
            local nx, ny = node.x, node.y

            table.insert(sprites, drawableSprite.fromTexture(
                ((i == 1 and not startOpen) or (i == #fullNodes and not endOpen)) and endTexture or pointTexture,
                { x = nx, y = ny, color = color }
            ))
        end

        return sprites
    end,

    nodeSprite = function () end,
}

return smwTrack
